php/base-stemmer.php

```diff
@@ -0,0 +1,524 @@
+        while ( $this->cursor > $this->limit_backward) {
+            $ch = $this->currentCharCodeAt($this->cursor - 1);
+            if ($ch > $max || $ch < $min){ 
+                return true;
+            }
+            $ch -= $min;
+            if( ($s[ self::unsignedShiftRight($ch,3) ] & (0x1 << ($ch & 0x7))) === 0){
+                return true;
+            }
+            $this->cursor--;
+        }
+        return false;
+    }
+
+
+    /**
+     * @param int[] $s
+     */
+    protected function out_grouping( array $s, int $min, int $max ):bool {
+        if ($this->cursor >= $this->limit){
+            return false;
+        }
+        $ch = $this->currentCharCodeAt($this->cursor);
+        if ($ch > $max || $ch < $min) {
+            $this->cursor++;
+            return true;
+        }
+        $ch -= $min;
+        if( ($s[ self::unsignedShiftRight($ch,3) ] & (0x1 << ($ch & 0x7))) === 0) {
+            $this->cursor++;
+            return true;
+        }
+        return false;
+    }
+
+
+    /**
+     * @param int[] $s
+     */
+    protected function go_out_grouping ( array $s, int $min, int $max ):bool {
+        while ( $this->cursor < $this->limit) {
+            $ch = $this->currentCharCodeAt($this->cursor);
+            if ($ch <= $max && $ch >= $min) {
+                $ch -= $min;
+                if( ($s[ self::unsignedShiftRight($ch,3) ] & (0x1 << ($ch & 0x7))) !== 0 ) {
+                    return true;
+                }
+            }
+            $this->cursor++;
+        }
+        return false;
+    }
+
+
+    /**
+     * @param int[] $s
+     */
+    protected function out_grouping_b( array $s, int $min, int $max ):bool {
+        if ($this->cursor <= $this->limit_backward){
+            return false;
+        }
+        $ch = $this->currentCharCodeAt($this->cursor - 1);
+        if ($ch > $max || $ch < $min) {
+            $this->cursor--;
+            return true;
+        }
+        $ch -= $min;
+        if( ($s[ self::unsignedShiftRight($ch,3) ] & (0x1 << ($ch & 0x7))) === 0 ) {
+            $this->cursor--;
+            return true;
+        }
+        return false;
+    }
+
+
+    /**
+     * @param int[] $s
+     */
+    protected function go_out_grouping_b ( array $s, int $min, int $max ):bool {
+        while ( $this->cursor > $this->limit_backward) {
+            $ch = $this->currentCharCodeAt($this->cursor - 1);
+            if ($ch <= $max && $ch >= $min) {
+                $ch -= $min;
+                if( ($s[ self::unsignedShiftRight($ch,3) ] & (0x1 << ($ch & 0x7))) !== 0) {
+                    return true;
+                }
+            }
+            $this->cursor--;
+        }
+        return false;
+    }
+
+
+    protected function eq_s( string $s ):bool {
+        $slength = self::lengthOf($s);
+        if ($this->limit - $this->cursor < $slength ){
+            return false;
+        }
+        if ( $this->currentSlice($this->cursor, $this->cursor + $slength) !== $s) {
```

Comment:
Presumably using `substr_compare()` would be more efficient here as it would avoid creating a temporary string (unless there's some funky lazy slicing going on):

https://www.php.net/manual/en/function.substr-compare.php